src/target/riscv/riscv.c

```diff
@@ -2697,7 +2768,158 @@ COMMAND_HANDLER(handle_repeat_read)
 	return result;
 }
 
+COMMAND_HANDLER(handle_memory_sample_command)
+{
+	struct target *target = get_current_target(CMD_CTX);
+
+	if (CMD_ARGC == 0) {
+		for (unsigned i = 0; i < DIM(sample_config.bucket); i++) {
+			if (sample_config.bucket[i].enabled) {
+				command_print(CMD, "bucket %d; address=0x%" TARGET_PRIxADDR "; size=%d", i,
+							  sample_config.bucket[i].address,
+							  sample_config.bucket[i].size_bytes);
+			} else {
+				command_print(CMD, "bucket %d; disabled", i);
+			}
+		}
+		return ERROR_OK;
+	}
+
+	if (CMD_ARGC < 2) {
+		LOG_ERROR("Command requires at least bucket and address arguments.");
+		return ERROR_FAIL;
+	}
+
+	if (riscv_rtos_enabled(target)) {
+		LOG_ERROR("Memory sampling is not supported with `-rtos riscv`.");
+		return ERROR_FAIL;
+	}
+
+	uint32_t bucket;
+	COMMAND_PARSE_NUMBER(u32, CMD_ARGV[0], bucket);
+	if (bucket > DIM(sample_config.bucket)) {
+		LOG_ERROR("Max bucket number is %d.", (unsigned) DIM(sample_config.bucket));
+		return ERROR_COMMAND_ARGUMENT_INVALID;
+	}
+
+	sample_config.bucket[bucket].enabled = true;
```

Comment:
How about disabling the memory sampling for the given buckets?

Or is it assumed that the user will configure the sampling once and will leave it that way for the whole OpenOCD session? If that is the case, then it would make sense to allow the `riscv memory_sample` command only in the configuration phase (i.e. `COMMAND_CONFIG` instead of `COMMAND_ANY`).

Reply:
Disabling is a good point. Added that.